torchdata/nodes/map.py

```diff
@@ -0,0 +1,132 @@
+        method: Literal["thread", "process"] = "thread",
+    ):
+        self.source = source
+        self.udf = map_fn
+        self.num_workers = num_workers
+        self.in_order = in_order
+
+        self.in_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.out_q: Union[queue.Queue, mp.Queue] = queue.Queue() if method == "thread" else mp.Queue()
+        self.sem = threading.BoundedSemaphore(value=2 * num_workers)
+
+        self._started = False
+        self._stop = threading.Event()
+        self._mp_stop = mp.Event()
+        self._read_thread = threading.Thread(
+            target=_populate_queue, args=(self.source, self.in_q, self._stop, self.sem)
+        )
+        self._method = method
+
+        self._map_threads: List[Union[threading.Thread, mp.Process]] = []
+
+    def iterator(self) -> Iterator[T]:
+        if not self._started:
+            for worker_id in range(self.num_workers):
+                args = (
+                    worker_id,
+                    self.in_q,
+                    self.out_q,
+                    self.in_order,
+                    self.udf,
+                    self._stop if self._method == "thread" else self._mp_stop,
+                )
+                self._map_threads.append(
+                    threading.Thread(target=_apply_udf, args=args)
+                    if self._method == "thread"
+                    else mp.Process(target=_apply_udf, args=args)
+                )
+            self._read_thread.start()
+            for t in self._map_threads:
+                t.start()
+            self._started = True
+
+        exception: Optional[ExceptionWrapper] = None
+        while True:
+            if self._stop.is_set():
+                yield from self._flush_queues()
+                self._mp_stop.set()
+                break
+            try:
+                item = self.out_q.get(block=True, timeout=1.0)
+                self.sem.release()
+            except queue.Empty:
+                continue
+            if isinstance(item, StopIteration):
+                yield from self._flush_queues()
+                break
+            elif isinstance(item, ExceptionWrapper):
+                exception = item
+                break
+            yield item
+
+        self._stop.set()
+        self._mp_stop.set()
+        if exception is not None:
+            exception.reraise()
+        self._shutdown()
+
+    def _flush_queues(self):
+        while self.sem._value < 2 * self.num_workers:
+            x = self.out_q.get(block=True, timeout=5.0)
+            self.sem.release()
+            yield x
+
+    def __del__(self):
+        self._shutdown()
+
+    def _shutdown(self):
+        if self._started:
+            self._stop.set()
+            self._mp_stop.set()
+            for _ in range(5):
+                self._read_thread.join(timeout=0.1)
+                if not self._read_thread.is_alive():
+                    break
```

Comment:
If we weren't able to successfully join with the thread after 5 tries, should we log something? I get that we don't want to block, and raising an exception would also be rude.